### PR TITLE
Store state versions

### PR DESCRIPTION
**Pull request by talekhinezh**
* Separate atom into two types: ClientAtom + CommittedAtom.
* The CommittedAtom contains ledger specific information (e.g. the stateVersion) whereas the ClientAtom does not
* The CommittedAtom is what gets stored persistently.

### CODE REVIEW

File: radixdlt/src/integration/java/com/radixdlt/consensus/tempo/RadixLedgerEntryTests.java [hunk 1]
```diff
@@ -43,15 +43,6 @@ public void store_atom() throws Exception {
 		// TODO should check LocalSystem clocks once implemented
 	}
 
-	@Test
```
Comment (flotothemoon) [inline radixdlt/src/integration/java/com/radixdlt/consensus/tempo/RadixLedgerEntryTests.java:46]: Why was this test removed?

File: radixdlt/src/integration/java/com/radixdlt/consensus/tempo/LedgerEntryGenerator.java [hunk 2]
```diff
@@ -32,6 +32,9 @@
 import java.util.Random;
 
 public class LedgerEntryGenerator {
+
```
Comment (flotothemoon) [inline radixdlt/src/integration/java/com/radixdlt/consensus/tempo/LedgerEntryGenerator.java]: Extra new line?

File: radixdlt/src/main/java/com/radixdlt/consensus/VertexStore.java [hunk 3]
```diff
@@ -69,10 +70,14 @@ public VertexStore(
 		this.engine = Objects.requireNonNull(engine);
 		this.counters = Objects.requireNonNull(counters);
 		this.highestQC = Objects.requireNonNull(rootQC);
-		try {
-			this.engine.checkAndStore(genesisVertex.getAtom());
-		} catch (RadixEngineException e) {
-			throw new IllegalStateException("Could not store genesis atom: " + genesisVertex.getAtom(), e);
+
+		if (genesisVertex.getAtom() != null) {
```
Comment (flotothemoon) [inline radixdlt/src/main/java/com/radixdlt/consensus/VertexStore.java:74]: Do we allow the genesis vertex not to have an atom now?

File: radixdlt/src/main/java/com/radixdlt/mempool/MempoolDuplicateException.java [hunk 4]
```diff
@@ -17,14 +17,14 @@
 
 package com.radixdlt.mempool;
 
-import com.radixdlt.middleware2.LedgerAtom;
+import com.radixdlt.middleware2.ClientAtom;
 
 /**
  * Exception thrown when an attempt to add new items would
```
Comment (flotothemoon) [inline radixdlt/src/main/java/com/radixdlt/mempool/MempoolDuplicateException.java]: Appreciate this isn't new, just realised this description doesn't match the exception name ("capacity" vs "duplicate")

File: radixdlt/src/main/java/com/radixdlt/middleware2/CommittedAtom.java [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+import java.util.Objects;
+import javax.annotation.concurrent.Immutable;
+
+/**
+ * An atom which has been committed by the BFT
+ *
+ * TODO: add commit signature proof
+ */
+@Immutable
+@SerializerId2("consensus.committed_ledger_atom")
```
Comment (flotothemoon) [inline radixdlt/src/main/java/com/radixdlt/middleware2/CommittedAtom.java]: "consensus.committed_atom" might be more succinct

**Review by flotothemoon — APPROVED** [review]
All good, some minor comments.

File: radixdlt/src/main/java/com/radixdlt/consensus/VertexStore.java (hunk 3 above)
Comment (talekhinezh) [reply radixdlt/src/main/java/com/radixdlt/consensus/VertexStore.java:74]: yes should be allowable

File: radixdlt/src/main/java/com/radixdlt/consensus/VertexStore.java [hunk 6]
```diff
@@ -69,10 +70,14 @@ public VertexStore(
...
+			CommittedAtom committedGenesis = genesisVertex.getAtom().committed(rootQC.getProposed());
+			try {
+				this.engine.checkAndStore(committedGenesis);
```
Comment (clstrfsck) [inline radixdlt/src/main/java/com/radixdlt/consensus/VertexStore.java:77]: Maybe now would be a good time to only store if not already present?

**Review by clstrfsck — DISMISSED** [review]
Looks fine, but maybe should consider how to deal with genesis atom unconditional checkAndStore now.

File: radixdlt/src/main/java/com/radixdlt/middleware2/CommittedAtom.java [hunk 7]
```diff
@@ -0,0 +1,109 @@
...
+		return clientAtom.getCMInstruction();
+	}
+
+	@Override
+	public AID getAID() {
+		return clientAtom.getAID();
+	}
+
+	@Override
+	public int hashCode() {
```
Comment (clstrfsck) [inline radixdlt/src/main/java/com/radixdlt/middleware2/CommittedAtom.java:85]: Should have EqualsVerifier tests for hashCode and equals here.

File: radixdlt/src/main/java/com/radixdlt/mempool/MempoolDuplicateException.java (hunk 4 above)
Comment (clstrfsck) [reply radixdlt/src/main/java/com/radixdlt/mempool/MempoolDuplicateException.java]: Looks like I made copypasta.

**Review by clstrfsck — COMMENTED** [review]
Test coverage seems a bit low in general.  Probably not worth investing significant effort in tests for the API stuff right now, but could improve in areas that are not currently scheduled for significant rework.

Really need an EqualsVerifier test for ComittedAtom as noted.

File: radixdlt/src/integration/java/com/radixdlt/consensus/tempo/RadixLedgerEntryTests.java (hunk 1 above)
Comment (talekhinezh) [reply radixdlt/src/integration/java/com/radixdlt/consensus/tempo/RadixLedgerEntryTests.java:46]: Don't accept duplicates anymore

File: radixdlt/src/main/java/com/radixdlt/consensus/VertexStore.java (hunk 6 above)
Comment (talekhinezh) [reply radixdlt/src/main/java/com/radixdlt/consensus/VertexStore.java:77]: I think this can be saved for later

**Pull request merged**